kratos/future/input_output/vtu_output.cpp

```diff
@@ -0,0 +1,1689 @@
+    const NDData<int>& rOffsets,
+    const TContainerType& rContainer,
+    const std::unordered_map<IndexType, IndexType>& rKratosVtuIndicesMap,
+    const std::vector<IndexType>& rWritingIndices)
+{
+    if (rOffsets.Size() == 0) {
+        return Kratos::make_shared<NDData<int>>(DenseVector<unsigned int>(1, 0));
+    }
+
+    const auto offsets_span = rOffsets.ViewData();
+    auto p_connectivities = Kratos::make_shared<NDData<int>>(DenseVector<unsigned int>(1, offsets_span.back()));
+    auto connectivities_span = p_connectivities->ViewData();
+
+    IndexPartition<IndexType>(rWritingIndices.size()).for_each([&connectivities_span, &offsets_span, &rContainer, &rKratosVtuIndicesMap, &rWritingIndices](const IndexType Index) {
+        const auto& r_geometry = (rContainer.begin() + rWritingIndices[Index])->GetGeometry();
+        auto entity_data_begin_itr = connectivities_span.begin() + offsets_span[Index] - r_geometry.size();
+
+        for (IndexType i_node = 0; i_node < r_geometry.size(); ++i_node) {
+            const auto p_itr = rKratosVtuIndicesMap.find(r_geometry[i_node].Id());
+            if (p_itr != rKratosVtuIndicesMap.end()) {
+                entity_data_begin_itr[i_node] = p_itr->second;
+            } else {
+                KRATOS_ERROR << "Node with id " << r_geometry[i_node].Id() << " not found in nodes list.";
+            }
+        }
+    });
+
+    return p_connectivities;
+}
+
+template<class TXmlDataElementWrapper>
+void AddConnectivityData(
+    XmlElementsArray& rCellElement,
+    std::vector<IndexType>& rWritingIndices,
+    const ModelPart::NodesContainerType& rNodes,
+    VtuOutput::CellContainerPointerType pCells,
+    TXmlDataElementWrapper& rXmlDataElementWrapper,
+    const IndexType EchoLevel)
+{
+    std::visit([&rCellElement, &rXmlDataElementWrapper, &rWritingIndices, &rNodes, EchoLevel](auto p_container) {
+        std::unordered_map<IndexType, IndexType> indices_map;
+        indices_map.reserve(rNodes.size());
+        IndexType vtu_index = 0;
+        for (const auto& r_node : rNodes) {
+            indices_map[r_node.Id()] = vtu_index++;
+        }
+
+        KRATOS_INFO_IF("VtuOutput", EchoLevel > 2) << "------ Collecting Vtk geometry type info...\n";
+        auto p_type_data = GetGeometryTypes(rWritingIndices, *p_container, EchoLevel);
+
+        KRATOS_INFO_IF("VtuOutput", EchoLevel > 2) << "------ Ignored " << (p_container->size() - rWritingIndices.size()) << "/"
+            << p_container->size() << " " << GetEntityName(p_container) << "(s).\n";
+
+        KRATOS_INFO_IF("VtuOutput", EchoLevel > 2) << "------ Collecting geometry offsets info...\n";
+        auto p_offsets = GetOffsets(rWritingIndices, *p_container);
+
+        KRATOS_INFO_IF("VtuOutput", EchoLevel > 2) << "------ Collecting geometry connectivity info...\n";
+        rCellElement.AddElement(rXmlDataElementWrapper.Get("connectivity", GetConnectivities(*p_offsets, *p_container, indices_map, rWritingIndices)));
```

Comment:
why using a map here within the XMLDataElementWrapper? wouldn't a struct be valid?

Reply:
`XmlDataelementWrapper` is a wrapper class which produces the correct `XmlElement` depending on the `writer_type` chosen by the user for given `NDData`. It does not have a map inside. May be i am not understanding what you mean by a map?

It needs to have `XmlDataElementWrapper::Get` method as virtual to have the dynamic branching depending on the user input for `writer_format`, being `ASCII, BINARY, RAW, COMPRESSED_RAW`.